src/cache/BrowserBasedPackageCache.ts

```diff
@@ -0,0 +1,230 @@
+import { Readable } from 'stream';
+import zlib from 'zlib';
+import tar from 'tar-stream';
+import { LogFunction } from '../utils';
+import { PackageCache, PackageCacheOptions } from './PackageCache';
+
+export class BrowserBasedPackageCache implements PackageCache {
+  private log: LogFunction;
+  private localResourceMap: Map<string, Map<string, any>>;
+
+  constructor(
+    private databaseName: string,
+    options: PackageCacheOptions = {}
+  ) {
+    this.log = options.log ?? (() => {});
+    this.localResourceMap = new Map();
+  }
+
+  // NOTE: need to call this after construction whenever using this PackageCache
+  async initialize(dependencies: { name: string; version: string }[]): Promise<void> {
+    const packageLabels = dependencies.map(d => `${d.name}#${d.version}`);
+    return new Promise(resolve => {
+      const openRequest = indexedDB.open(this.databaseName);
+      openRequest.onsuccess = event => {
+        const database = (event.target as IDBOpenDBRequest).result;
+        const initializePromises = packageLabels.map(packageLabel =>
+          initializeResourcesInLocalMap(database, packageLabel)
+        );
+        Promise.allSettled(initializePromises)
+          .then(results =>
+            results
+              .filter(result => result.status === 'fulfilled')
+              .forEach(result => {
+                const [packageLabel, resources] = result.value;
+                this.localResourceMap.set(packageLabel, resources);
+              })
+          )
+          .then(() => {
+            database.close();
+            resolve();
+          });
+      };
+    });
+  }
+
+  async cachePackageTarball(name: string, version: string, data: Readable): Promise<string> {
+    const packageLabel = `${name}#${version}`;
+
+    if (version === 'current' || version === 'dev') {
+      // not currently supported in browsers
+      throw Error('"current" or "dev" package versions are not yet supported.');
+    }
+
+    return new Promise(resolve => {
+      const resources: any[] = [];
+      // Extract the package
+      const extract = tar.extract();
+      extract.on('entry', (header, stream, next) => {
+        let resourceData = '';
+        stream.on('data', chunk => {
+          resourceData += chunk.toString();
+        });
+        stream.on('end', () => {
+          try {
+            // Only add:
+            // - files that are directly in 'packages' and not in nested folders to align with DiskBasedPackageCache, which reads directories non-recursively
```

Comment:
Minor typo: the string `package` in the code is correct, so the comment should be the same.

```suggestion
            // - files that are directly in 'package' and not in nested folders to align with DiskBasedPackageCache, which reads directories non-recursively
```

Reply:
Doh. Thanks!